### PR TITLE
Autologging: Fix bug where metrics are not logged for single-epoch tf.keras training sessions

### PR DESCRIPTION
**Pull request by dbczumar**
## What changes are proposed in this pull request?

This PR fixes an issue with `tf.keras` autologging where, if a model is trained for exactly 1 epoch, metrics are not recorded. This can be reproduced by running the following code and observing a lack of expected metrics (`loss` and `accuracy`) in the resulting `mlruns` directory:

```
import mlflow

mlflow.autolog()

import tensorflow as tf

mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

model = tf.keras.models.Sequential(
    [
        tf.keras.layers.Flatten(input_shape=(28, 28)),
        tf.keras.layers.Dense(128, activation="relu"),
        tf.keras.layers.Dropout(0.2),
        tf.keras.layers.Dense(10),
    ]
)

model.compile(
    optimizer="adam",
    loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
    metrics=["accuracy"],
)
model.fit(x_train, y_train, epochs=1)
```

This PR fixes the issue and introduces test coverage to ensure that it does not recur in tf.keras and does not arise for other TensorFlow metrics APIs.

## How is this patch tested?

Integration tests

## Release Notes

Fix an autologging bug where tf.keras models trained for a single epoch did not log metrics.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [X] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/tensorflow.py [hunk 1]
```diff
@@ -615,7 +615,12 @@ def _log_event(event):
         summary = event.summary
         for v in summary.value:
             if v.HasField("simple_value"):
-                if (event.step - 1) % _LOG_EVERY_N_STEPS == 0:
+                # In tf.keras, epoch indexing for single-epoch training sessions begins at 0,
+                # while indexing begins at 1 for multi-epoch training sessions. While we have
+                # not observed similar discrepancies for other TensorFlow APIs, we cautiously
+                # insert some `max` logic to ensure that metrics for the first epoch are logged,
+                # whether that epoch is referred to as `0` or `1`
+                if max(event.step - 1, 0) % _LOG_EVERY_N_STEPS == 0:
```
Comment (mohamad-arabi) [inline mlflow/tensorflow.py:623]: Good catch! thanks for the clarification in comments

**Review by mohamad-arabi — APPROVED** [review]
LGTM

**Review by mohamad-arabi — APPROVED** [review]
LGTM!
